Sources/StreamVideo/CallStateMachine/Stages/Call+JoiningStage.swift

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright © 2025 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+extension Call.StateMachine.Stage {
+
+    /// Creates a joining stage for a call with the given input.
+    ///
+    /// - Parameters:
+    ///   - call: The call to be joined
+    ///   - input: The input containing join parameters
+    /// - Returns: A new `JoiningStage` instance
+    static func joining(
+        _ call: Call,
+        input: Context.JoinInput
+    ) -> Call.StateMachine.Stage {
+        JoiningStage(
+            .init(
+                call: call,
+                joinInput: input
+            )
+        )
+    }
+}
+
+extension Call.StateMachine.Stage {
+
+    /// Represents the joining stage in the call state machine.
+    final class JoiningStage: Call.StateMachine.Stage, @unchecked Sendable {
+        @Injected(\.streamVideo) private var streamVideo
+
+        private let disposableBag = DisposableBag()
+        private var iterations = 0
+
+        /// Creates a new joining stage with the provided context.
+        ///
+        /// - Parameter context: The call context containing necessary state
+        init(
+            _ context: Context
+        ) {
+            super.init(id: .joining, context: context)
+        }
+
+        /// Handles state transitions to the joining stage.
+        ///
+        /// Valid transitions:
+        /// - From: IdleStage, AcceptedStage, JoiningStage
+        /// - To: JoinedStage, ErrorStage
+        ///
+        /// - Parameter previousStage: The stage transitioning from
+        /// - Returns: Self if transition is valid, nil otherwise
+        override func transition(
+            from previousStage: Call.StateMachine.Stage
+        ) -> Self? {
+            switch previousStage.id {
+            case .idle, .accepted, .joining:
+                execute()
```

Comment:
So we start a new task, and we immediately return the stage. Can you explain a bit more why we do it like this?

Reply:
This is how our StateMachine works. We do that so that we asynchronously make any operation we need while we allow the state machine to transition to the new stage. That ensures that all operations to the StateMachine are synchronous and short running.